Proposal: add `sourceContext` property to the messages generated by the `eslint` object

**Context:**
We would like to port the JSCS default formatter over to ESLint. That formatter prints out the lines around the location of an error/warning. See #5860 for an example output.

**Problem**
Unfortunately, with ESLint, there is currently no way to get those lines of "context" from a formatter, because:

1. The formatter is invoked with only one argument - the array of messages to report on, and those do not contain any information about the lines of context around the error.
2. The formatter is intentionally made to be "dumb" and shouldn't/cannot use `SourceCode.getLines()` to retrieve the lines of code of the file being reported on.

**Proposal**
I would like to add a `sourceContext` property to each message object that gets returned by the `eslint.verify()` method.
With this new property, the message object would look roughly like the following:

```javascript
{
    ruleId: 'lines-around-comment',
    severity: 2,
    message: 'Expected line before comment.',
    line: 37,
    column: 9,
    nodeType: 'Line',
    source: '// some random comment',
    sourceContext: {
        before: [
            "line of code before line of the error",
            "another line of code"
        ],
        after: [
            "line of code after the line of the error",
            "another line of code"
        ]
    }
}
```

Each item of the `before`/`after` array is a line of code directly from the array returned by `SourceCode.getLines(code)`.

An alternative would be that each item is instead an object that contains a `source` and a `line` property, like this:

```javascript
{
    sourceContext: {
        before: [
            {
                source: "line of code before line of the error",
                line: 35
            },
            {
                source: "another line of code",
                line: 36
            }
        ],
        after: [
            // ...
        ]
    }
}
```
The advantage of this alternative approach is that it would prevent delegating the calculation of the line number to the formatter.

Adding this `sourceContext` property would solve the problem described, and would allow us to port the JSCS formatter with ease.
I have some proof-of-concept code for this proposal locally that is already able to print out the lines around an error, like this:

```bash
$ bin/eslint.js --rulesdir lib/internal-rules/ lib/formatters/stylish.js
35
36     results.forEach(function(result) {
37         // console.log(result);
38         const messages = result.messages;
39

lib/formatters/stylish.js
  37:9  error  Expected line before comment  lines-around-comment
```

The existing tests also already pass so I would only need to clean up the code a bit and make sure to test some edge cases better.

**Questions**
- Are there other any other ways to solve the problem described that would be preferable?
- I don't really like the property name `sourceContext` that much, but couldn't come up with anything better. Any suggestions?

--- a/docs/developer-guide/nodejs-api.md
+++ b/docs/developer-guide/nodejs-api.md
@@ -240,7 +240,15 @@ The return value is an object containing the results of the linting operation. H
 }
 ```
 
-The top-level report object has a `results` array containing all linting results for files that had warnings or errors (any files that did not produce a warning or error are omitted). Each file result includes the `filePath`, a `messages` array, `errorCount`, `warningCount`, and optionally `output`. The `messages` array contains the result of calling `linter.verify()` on the given file. The `errorCount` and `warningCount` give the exact number of errors and warnings respectively on the given file. The `output` property gives the source code for the file with as many fixes applied as possible, so you can use that to rewrite the files if necessary. The top-level report object also has `errorCount` and `warningCount` which give the exact number of errors and warnings respectively on all the files.
+The top-level report object has a `results` array containing all linting results for files that had warnings or errors (any files that did not produce a warning or error are omitted). Each file result includes:
+
+* `filePath` - Path to the given file.
+* `messages` - Array containing the result of calling `linter.verify()` on the given file.
+* `errorCount` and `warningCount` - The exact number of errors and warnings respectively on the given file.
+* `source` - The source code for the given file. This property is omitted if this file has no errors/warnings or if the `output` property is present.
+* `output` - The source code for the given file with as many fixes applied as possible, so you can use that to rewrite the files if necessary. This property is omitted if no fix is available.
+
+The top-level report object also has `errorCount` and `warningCount` which give the exact number of errors and warnings respectively on all the files.
 
 Once you get a report object, it's up to you to determine how to output the results. Fixes will not be automatically applied to the files, even if you set `fix: true` when constructing the `CLIEngine` instance. To apply fixes to the files, call [`outputFixes`](#outputfixes).
 

--- a/docs/developer-guide/nodejs-api.md
+++ b/docs/developer-guide/nodejs-api.md
@@ -124,6 +124,8 @@ The information available for each linting message is:
 * `endLine` - the end line of the range on which the error occurred (this property is omitted if it's not range).
 * `fix` - an object describing the fix for the problem (this property is omitted if no fix is available).
 
+**Please note**: the `source` property will be removed from the linting messages in an upcoming breaking release. If you depend on this property, you can still use the `getSourceCode` method described below to get the line of code for each message.
+
 You can also get an instance of the `SourceCode` object used inside of `linter` by using the `getSourceCode()` method:
 
 ```js
@@ -249,6 +251,8 @@ The top-level report object has a `results` array containing all linting results
 * `output` - The source code for the given file with as many fixes applied as possible, so you can use that to rewrite the files if necessary. This property is omitted if no fix is available.
 
 The top-level report object also has `errorCount` and `warningCount` which give the exact number of errors and warnings respectively on all the files.
+
+**Please note**: the `source` property will be removed from the linting messages returned in `messages` in an upcoming breaking release. If you depend on this property, you should now use the top-level `source` or `output` properties instead.
 
 Once you get a report object, it's up to you to determine how to output the results. Fixes will not be automatically applied to the files, even if you set `fix: true` when constructing the `CLIEngine` instance. To apply fixes to the files, call [`outputFixes`](#outputfixes).
 

--- a/docs/developer-guide/working-with-custom-formatters.md
+++ b/docs/developer-guide/working-with-custom-formatters.md
@@ -84,6 +84,8 @@ The following are the fields of the result object:
 - **nodeType**: the type of the node in the [AST](https://github.com/estree/estree/blob/master/spec.md#node-objects)
 - **source**: a extract of the code the line where the failure happened.
 
+**Please note**: the `source` property will be removed from the message object in an upcoming breaking release. If you depend on this property, you should now use the `source` or `output` properties from [the result object](#the-result-object) instead.
+
 ## Examples
 
 ### Summary formatter

--- a/docs/developer-guide/working-with-custom-formatters.md
+++ b/docs/developer-guide/working-with-custom-formatters.md
@@ -28,23 +28,24 @@ The output of the previous command will be something like this
                 "ruleId": "curly",
                 "severity": 2,
                 "message": "Expected { after 'if' condition.",
-                "line": 41,
-                "column": 2,
+                "line": 2,
+                "column": 1,
                 "nodeType": "IfStatement",
-                "source": "  if ( err ) console.log( 'failed tests: ' + err );"
+                "source": "if (err) console.log('failed tests: ' + err);"
             },
             {
                 "ruleId": "no-process-exit",
                 "severity": 2,
                 "message": "Don't use process.exit(); throw an error instead.",
-                "line": 42,
-                "column": 2,
+                "line": 3,
+                "column": 1,
                 "nodeType": "CallExpression",
-                "source": "  process.exit( exitCode );"
+                "source": "process.exit(1);"
             }
         ],
         "errorCount": 2,
-        "warningCount": 0
+        "warningCount": 0,
+        "source": "var err = doStuff();\nif (err) console.log('failed tests: ' + err);\nprocess.exit(1);\n"
     },
     {
         "filePath": "Gruntfile.js",
@@ -67,9 +68,11 @@ the list of messages for `errors` and/or `warnings`.
 The following are the fields of the result object:
 
 - **filePath**: The path to the file relative to the current working directory (the path from which eslint was executed).
-- **messages**: An array of message objects. See below for more info about messages
-- **errorCount**: The number of errors for the given file
-- **warningCount**: the number of warnings for the give file
+- **messages**: An array of message objects. See below for more info about messages.
+- **errorCount**: The number of errors for the given file.
+- **warningCount**: The number of warnings for the given file.
+- **source**: The source code for the given file. This property is omitted if this file has no errors/warnings or if the `output` property is present.
+- **output**: The source code for the given file with as many fixes applied as possible. This property is omitted if no fix is available.
 
 ### The message object
 

--- a/lib/cli-engine.js
+++ b/lib/cli-engine.js
@@ -75,6 +75,8 @@ const debug = require("debug")("eslint:cli-engine");
  * @property {LintMessage[]} messages All of the messages for the result.
  * @property {number} errorCount Number or errors for the result.
  * @property {number} warningCount Number or warnings for the result.
+ * @property {string=} [source] The source code of the file that was linted.
+ * @property {string=} [output] The source code of the file that was linted, with as many fixes applied as possible.
  */
 
 //------------------------------------------------------------------------------
@@ -175,7 +177,7 @@ function multipassFix(text, config, options) {
 
 
     /*
-     * If the last result had fixes, we need to lint again to me sure we have
+     * If the last result had fixes, we need to lint again to be sure we have
      * the most up-to-date information.
      */
     if (fixedResult.fixed) {
@@ -198,7 +200,7 @@ function multipassFix(text, config, options) {
  * @param {string} filename An optional string representing the texts filename.
  * @param {boolean} fix Indicates if fixes should be processed.
  * @param {boolean} allowInlineConfig Allow/ignore comments that change config.
- * @returns {Result} The results for linting on this text.
+ * @returns {LintResult} The results for linting on this text.
  * @private
  */
 function processText(text, configHelper, filename, fix, allowInlineConfig) {
@@ -279,6 +281,10 @@ function processText(text, configHelper, filename, fix, allowInlineConfig) {
         result.output = fixedResult.output;
     }
 
+    if (result.errorCount + result.warningCount > 0 && typeof result.output === "undefined") {
+        result.source = text;
+    }
+
     return result;
 }
 
@@ -288,7 +294,7 @@ function processText(text, configHelper, filename, fix, allowInlineConfig) {
  * @param {string} filename The filename of the file being checked.
  * @param {Object} configHelper The configuration options for ESLint.
  * @param {Object} options The CLIEngine options object.
- * @returns {Result} The results for linting on this file.
+ * @returns {LintResult} The results for linting on this file.
  * @private
  */
 function processFile(filename, configHelper, options) {
@@ -304,7 +310,7 @@ function processFile(filename, configHelper, options) {
  * Returns result with warning by ignore settings
  * @param {string} filePath - File path of checked code
  * @param {string} baseDir  - Absolute path of base directory
- * @returns {Result}           Result with single warning
+ * @returns {LintResult} Result with single warning
  * @private
  */
 function createIgnoreResult(filePath, baseDir) {
@@ -524,12 +530,13 @@ CLIEngine.getErrorResults = function(results) {
         const filteredMessages = result.messages.filter(isErrorMessage);
 
         if (filteredMessages.length > 0) {
-            filtered.push({
-                filePath: result.filePath,
-                messages: filteredMessages,
-                errorCount: filteredMessages.length,
-                warningCount: 0
-            });
+            filtered.push(
+                Object.assign(result, {
+                    messages: filteredMessages,
+                    errorCount: filteredMessages.length,
+                    warningCount: 0
+                })
+            );
         }
     });
 

--- a/tests/lib/cli-engine.js
+++ b/tests/lib/cli-engine.js
@@ -274,7 +274,8 @@ describe("CLIEngine", function() {
                             }
                         ],
                         errorCount: 1,
-                        warningCount: 0
+                        warningCount: 0,
+                        source: "var bar = foo"
                     }
                 ],
                 errorCount: 1,
@@ -349,7 +350,92 @@ describe("CLIEngine", function() {
                             }
                         ],
                         errorCount: 1,
-                        warningCount: 0
+                        warningCount: 0,
+                        source: "var bar ="
+                    }
+                ],
+                errorCount: 1,
+                warningCount: 0
+            });
+        });
+
+        it("should return source code of file in `source` property when errors are present", function() {
+            engine = new CLIEngine({
+                useEslintrc: false,
+                rules: { semi: 2 }
+            });
+
+            const report = engine.executeOnText("var foo = 'bar'");
+
+            assert.equal(report.results[0].source, "var foo = 'bar'");
+        });
+
+        it("should return source code of file in `source` property when warnings are present", function() {
+            engine = new CLIEngine({
+                useEslintrc: false,
+                rules: { semi: 1 }
+            });
+
+            const report = engine.executeOnText("var foo = 'bar'");
+
+            assert.equal(report.results[0].source, "var foo = 'bar'");
+        });
+
+
+        it("should not return a `source` property when no errors or warnings are present", function() {
+            engine = new CLIEngine({
+                useEslintrc: false,
+                rules: { semi: 2 }
+            });
+
+            const report = engine.executeOnText("var foo = 'bar';");
+
+            assert.lengthOf(report.results[0].messages, 0);
+            assert.isUndefined(report.results[0].source);
+        });
+
+        it("should not return a `source` property when fixes are applied", function() {
+            engine = new CLIEngine({
+                useEslintrc: false,
+                fix: true,
+                rules: {
+                    semi: 2,
+                    "no-unused-vars": 2
+                }
+            });
+
+            const report = engine.executeOnText("var msg = 'hi' + foo\n");
+
+            assert.isUndefined(report.results[0].source);
+            assert.equal(report.results[0].output, "var msg = 'hi' + foo;\n");
+        });
+
+        it("should return a `source` property when a parsing error has occurred", function() {
+            engine = new CLIEngine({
+                useEslintrc: false,
+                rules: { semi: 2 }
+            });
+
+            const report = engine.executeOnText("var bar = foothis is a syntax error.\n return bar;");
+
+            assert.deepEqual(report, {
+                results: [
+                    {
+                        filePath: "<text>",
+                        messages: [
+                            {
+                                ruleId: null,
+                                fatal: true,
+                                severity: 2,
+                                message: "Parsing error: Unexpected token is",
+                                line: 1,
+                                column: 19,
+                                source: "var bar = foothis is a syntax error."
+                            }
+                        ],
+                        errorCount: 1,
+                        warningCount: 0,
+                        source: "var bar = foothis is a syntax error.\n return bar;"
                     }
                 ],
                 errorCount: 1,
@@ -2435,6 +2521,39 @@ describe("CLIEngine", function() {
             assert.equal(report.warningCount, 1);
             assert.equal(report.results[0].errorCount, 0);
             assert.equal(report.results[0].warningCount, 1);
+        });
+
+        it("should return source code of file in the `source` property", function() {
+            process.chdir(originalDir);
+            const engine = new CLIEngine({
+                useEslintrc: false,
+                rules: { quotes: [2, "double"] }
+            });
+
+
+            const report = engine.executeOnText("var foo = 'bar';");
+            const errorResults = CLIEngine.getErrorResults(report.results);
+
+            assert.lengthOf(errorResults[0].messages, 1);
+            assert.equal(errorResults[0].source, "var foo = 'bar';");
+        });
+
+        it("should contain `output` property after fixes", function() {
+            process.chdir(originalDir);
+            const engine = new CLIEngine({
+                useEslintrc: false,
+                fix: true,
+                rules: {
+                    semi: 2,
+                    "no-console": 2
+                }
+            });
+
+            const report = engine.executeOnText("console.log('foo')");
+            const errorResults = CLIEngine.getErrorResults(report.results);
+
+            assert.lengthOf(errorResults[0].messages, 1);
+            assert.equal(errorResults[0].output, "console.log('foo');");
         });
     });
 

